### ISSUE TITLE
not using GPU via MLJ interface

### ISSUE THREAD

**Issue by Moelf** [Evovest/EvoTrees.jl#273]
It looks like the following is not utilizing GPU if called from MLJ, is this a known issue?

```julia
EvoTreeClassifier(;
        device="gpu",
        loss=:logloss,
        eta=0.1,
        max_depth=5,
        nrounds=30,
        )
```

**Comment by jeremiedb** [comment]
It's actually expected not to work as `device` isn't part of the supported kwarg of `EvoTreeClassifier`: https://evovest.github.io/EvoTrees.jl/dev/models/#EvoTreeClassifier

```julia
julia> config = EvoTreeClassifier(                                                                                                                                                                            
           max_depth = 4,                                                                                                                                                                                     
           device=:gpu                                                                                                                                                                                        
       )
EvoTreeClassifier(
  nrounds = 100,
  L2 = 0.0,
  lambda = 0.0,
  gamma = 0.0,
  eta = 0.1,
  max_depth = 4,
  min_weight = 1.0,
  rowsample = 1.0,
  colsample = 1.0,
  nbins = 64,
  alpha = 0.5,
  tree_type = "binary",
  rng = Random.MersenneTwister(123))
```

To `device` argument must instead be passed to the `fit_evotree` or inference function: https://evovest.github.io/EvoTrees.jl/dev/#GPU-Acceleration: 
```julia
m = fit_evotree(config, dtrain; target_name="y", device="gpu");
```

I'm not clear if there would be an easy way to handle this within MLJ, since as discussed [here](https://discourse.julialang.org/t/mlj-on-gpu/55047/2), the approach in MLJ has been to specify accelerator through a model paramter, rather than at "runtime", when the `fit` method is called. The choice made in EvoTrees was done to reflect the nature of our workflows where we want to generate lists of hyper-parameters for hyper-search process and didn't want to tie these hyper-params up-front to the device on which our the models would be trained. 
I recognize there's a a bit of a philisophical question on whether some configurations like `device` / `accelarator`, or other iterative-specific mechanisms like `early_stopping_rounds` or `metric` should be specified as model hyper-parameters, or rather as configuration of the fitting step. I've been leaning on the latter; but would be happy to hear your thoughts.

**Comment by Moelf** [comment]
I guess that's a design choice / issue for @ablaom to answer. Specifically, for models such as `MLJFlux`, you specify a keyword for model such as `acceleration=...`, so I guess the expectation here is we need accept and keep-track of `acceleration...` inside `EvoTreeClassifier`, which is a bit annoying.

Second point is that the constructor of `EvoTreeClassifier` should probably validate the keywords arguments?

**Comment by ablaom** [comment]
This is something I've given some thought to. While this design choice at MLJ is unlikely to change, I'd be happy to hear dissenting views in light of our work at LearnAPI.jl.

I feel that for simplicity it's hard to justify model-specific instructions that are just for run-time. For example,  in such a design, things get complicated when your model is part of composite, such as a in a pipeline. In that case there's no simple interface point for passing a runtime parameter to a *single* component of the composite. And device type is not something you want to specify *uniformly* for all components of a composite, even if they all support it. 

As far as the point about restricting config parameters to those you may want to optimise, another approach would be to provide a trait that flags the parameters that can be optimised, or even better, as caret does, provide, for each such parameter,  a default prior for a random search.

**Comment by jeremiedb** [comment]
@Moelf Latest release of EvoTrees (v0.16.9) should resolved #274 and it now includes an `@info` when extra kwargs are passed to the model/learner constructor (`EvoTreeClassifier`, `EvoTreeRegressor`) which are ignored by it. 

Regarding the model specifications; I think that for the sake of enabling GPU support through MLJ, it would be worth that I reconsider an alignement of EvoTrees/NeuroTreeModels constructors with MLJ expectations. Hence the following kwargs from the `fit` / `fit_evotree` function would be moved into the contructors: 
- `device` / `gpuID`
- `metric`
- `early_stopping_rounds`

As well as variables names roles for when used with the Tables based internal API:
- `fnames`  / `feature_names`
- `target_name`
- `weight_name` / `weight_name`

It could make for a quick `v0.17.0` breaking release.

**Comment by jeremiedb** [comment]
@Moelf I've brought some of the changes mentionned above, which should make training on GPU through MLJ work: https://github.com/Evovest/EvoTrees.jl/pull/276
The `device` kwarg is now passed to the model contructor. I'd be happy to hear if it effectively solves the issue you encountered.

I've let the reference to variable names like `target_names` & cie untouched, so they remain kwrags to `fit_evotree` (not related to MLJ training loop anyway). 

Note that performance on GPU is good only for large datasets; and declined when getting into larges depths (ex: 10+). 
I have WIP on an alternative design to get closer to linear scaling with depth of trees, but there's still work needed as there seems to be performance trade-off between the 2 designs but would prefer to maintain a single one.

**Comment by ablaom** [comment]
Perhaps I mentioned this elsewhere, but I wanted to note that MLJ uses ComputationalResources for `acceleration` options. So, for example, instead of `acceleration=:gpu` you do the more specific `acceleration=CUDALibs()`. [docs](https://juliaai.github.io/MLJ.jl/dev/acceleration_and_parallelism/)

**Comment by jeremiedb** [comment]
Thanks Anthony. Yes I recall you pointed me to this `acceleration` API before. 
An aspect that I attempt to balance is the ability to specify a contructor/learner from non-Julia specific structs: so essentially Float/Int/String/Symbol. This is notably influenced by dealing with our internal hyper-search workflows where we generate grid of configs and it's been preferred to have these expressed as simple json.  

Also, my reading was that ComputationalResources.jl seemed stalled, and that other initiatives like Lux's https://github.com/LuxDL/Lux.jl/tree/main/lib/MLDataDevices may be more actual. 

Another consideration is that from my understanding these device pointers don't handle how to specify specific host, like GPU id when multiple gpus are present.

In that context, I felt more comfortable sticking to the original "cpu"/"gpu" arguments, which is aligned with the XGBoost/CatBoost approach.

**Comment by ablaom** [comment]
@jeremiedb These are all very good reasons for the current choice.

Thanks for pointing out MLDataDevices.

**Comment by jeremiedb** [comment]
@Moelf I'm closing as it the issue should now be resolved. With EvoTrees v0.17, the following constructor should now properly allow for GPU training with MLJ:

```julia
learner = EvoTreeClassifier(;
    device=:gpu,
    loss=:logloss,
    eta=0.1,
    max_depth=5,
    nrounds=30,
)
```
